### PR TITLE
chore: refactor CSV bulk upload

### PR DESCRIPTION
**Pull request by fzhao99**
# 🔀 PULL REQUEST

## 💡 Summary
As part of #1160 started refactoring the CSV bulk upload route. Split that out into a follow up here that refactors the bulk upload process to better follow code base patterns

## 🔗 Related Issue

Fixes #1419
- #1419

## ✅ Acceptance Criteria
- [ ] CSV bulk upload code is more cleanly factored and maintainable
- [ ] More tests exist for the edge case rendering cases
- [ ] No regressions throughout the experience

### CODE REVIEW

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx [hunk 1]
```diff
@@ -190,42 +136,38 @@ export function ImportCustomCodes({
       },
       {
         onSuccess: (res) => {
-          const preview = res.data.preview ?? [];
-          const previewCodeSystems = res.data.code_systems ?? {};
+          setUploadRowErrors(null);
 
-          if (preview.length === 0) {
```
Comment (fzhao99) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]: we handle the case where there aren't preview items [on the server](https://github.com/CDCgov/dibbs-ecr-refiner/pull/1310/changes#diff-020578f48b03d3ecd9922d46784868de6be058e05f7a4408a0bbc644c4bbb495), so opting to remove this check

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+    render(
+      <MemoryRouter
+        initialEntries={[`/configurations/${MOCK_CONFIG_ID}/build`]}
+      >
+        <TestQueryClientProvider>
+          <ImportCustomCodes configurationId={MOCK_CONFIG_ID} />
+        </TestQueryClientProvider>
+      </MemoryRouter>
+    );
+    expect(await screen.findByText('Import from CSV')).toBeInTheDocument();
```
Comment (fzhao99) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx]: Tried my best to scope the tests in this file to the front-end state related functions related to the preview steps and delegate the other round-trip functions to the e2e tests, but if folks have an edge case I've missed or think I should put something elsewhere, let me know

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+    await userEvent.type(searchInput, MOCK_OTHER_CODE_SUFFIX);
+    checkOtherCode();
+    checkCvxCode(false);
+    checkLoincCode(false);
+    checkSnomedCode(false);
+    checkIcd10Code(false);
+  });
+});
+
+function checkLoincCode(exists = true) {
```
Comment (fzhao99) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx:251]: open to deleting these wrappers if we think they're redundant, but think they make the tests a little easier to read rather than passing in the strings themselves

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx [hunk 4]
```diff
@@ -115,9 +114,9 @@ describe('Custom codes upload', () => {
     checkAllCodesExistence();
 
     await userEvent.click(
-      await within(deleteRows[0]).findByRole('button', { name: 'Delete' })
+      await within(deleteRows[1]).findByRole('button', { name: 'Delete' })
```
Comment (fzhao99) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx:123]: as part of this refactor, added [a screenreader-only header](https://github.com/CDCgov/dibbs-ecr-refiner/pull/1310/changes#diff-3dcfba97458016a65f4a1b5b47ee540a5e1381bd40e5a5a3115a2792c747429eR406) to the custom codes table, which bumps these row getters up by one index value

File: client/e2e/configuration.spec.ts [hunk 5]
```diff
@@ -503,8 +503,10 @@ test.describe('Configuration detail flow', () => {
       await expect(page.getByText('Row 2', { exact: false })).toBeVisible();
       await expect(
         page.getByText(
-          'Invalid system_key: ICD-1. [code_system] must be one of',
-          { exact: false }
+          'Invalid system: ICD-1. [code_system] must be one of [SNOMED, LOINC, ICD-10, RxNorm, CVX, Other]',
+          {
+            exact: false,
```
Comment (jakewheeler) [inline client/e2e/configuration.spec.ts]:
```suggestion
            exact: true,
```

File: client/e2e/configuration.spec.ts [hunk 6]
```diff
@@ -487,7 +487,7 @@ test.describe('Configuration detail flow', () => {
 
       await expect(
         page.getByRole('alert').filter({
-          hasText: /^CSV must contain headers: code,code_system,display_name$/,
+          hasText: /^CSV must contain headers: code, code_system, display_name/,
```
Comment (jakewheeler) [inline client/e2e/configuration.spec.ts:490]: I manually tested this and noticed that when I upload a CSV with incorrect headers I'm unable to proceed with any other actions. The screen seems to get permanently stuck in this error state and I need to navigate away and return before I can try again.

Your work here didn't add this FYI. I'm seeing it in main as well. Not sure if we already knew about this but figured I'd mention it in case we need a ticket for it.

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx (hunk 3 above)
Comment (jakewheeler) [reply client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx:251]: I considered this in your other PR but ultimately felt it's fine if we want to keep them around.

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx [hunk 7]
```diff
@@ -15,37 +13,38 @@ import { ConfirmModal, UndoModal, PreviewEditModal } from './Modals';
...
-type PreviewRow = {
-  item: SearchPreviewItem;
-  matches?: readonly FuseResultMatch[];
-};
+type RowError = { row: number; error: string };
+
+type OpenModalType = 'confirm' | 'undo' | 'preview' | 'none';
 
-interface ImportCustomCodesProps {
+type ImportCustomCodesProps = {
```
Comment (jakewheeler) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]: I wouldn't mind generally sticking with `interface` as our default unless we have a situation in which `type` is the obvious choice to reach for (such as `OpenModalType` right above). Any opinion on this?

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx [hunk 8]
```diff
@@ -15,37 +13,38 @@ import { ConfirmModal, UndoModal, PreviewEditModal } from './Modals';
...
+  };
 };
 
-type PreviewRow = {
-  item: SearchPreviewItem;
-  matches?: readonly FuseResultMatch[];
-};
+type RowError = { row: number; error: string };
+
+type OpenModalType = 'confirm' | 'undo' | 'preview' | 'none';
```
Comment (jakewheeler) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]: Maybe we call this `OpenModalState`?

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx [hunk 9]
```diff
@@ -56,126 +55,82 @@ export function ImportCustomCodes({
...
-  }, [step]);
+  const [error, setError] = useState<string | null>(null);
+  const [uploadRowErrors, setUploadRowErrors] = useState<RowError[] | null>(
+    null
+  );
 
-  const uploading = isUploading || isUploadPending;
+  const [step, setStep] = useState<CsvImportStep>('intro');
+  const [isUploading, setIsUploading] = useState(false);
+  const [curOpenModal, setCurOpenModal] = useState<OpenModalType>('none');
```
Comment (jakewheeler) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]: Did you research at all what it could look like to switch over to using `useReducer`? I don't normally reach for it but it might help in this scenario since we are using 8 `useState`s 🙈

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx [hunk 10]
```diff
@@ -472,7 +322,11 @@ export function ImportCustomCodes({
           </div>
         ) : step === 'intro' ? (
           <div className="w-full max-w-xl space-y-6">
-            {renderCsvInstructions()}
+            <UploadInstructions
```
Comment (jakewheeler) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx:294]: Thanks 🙏

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx [hunk 11]
```diff
@@ -554,51 +411,26 @@ export function ImportCustomCodes({
...
             </div>
             <table className="w-full border-separate border-spacing-y-2 text-left text-sm">
+              <thead className="sr-only">
+                <tr>
+                  <th>Custom code</th>
+                  <th>Custom code system</th>
+                  <th>Custom code name</th>
+                  <th>Modify the custom code</th>
+                </tr>
+              </thead>
```
Comment (jakewheeler) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]: Good catch 😬

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx (hunk 4 above)
Comment (jakewheeler) [reply client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx:123]: It'd be great to make this visible to everyone when we go to change the other custom codes table.

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx (hunk 9 above)
Comment (fzhao99) [reply client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]: hmm, it's a thought I had! I think in this particular case I'm struggling to see what the right abstraction to centralize things into a single reducer function would look like. But taking a look at this again, I think I may be able to compress these down even further if I componentize some of the individual items a bit. 

I'll hack on it a bit and let you know when I'm done.

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/Modals.tsx [hunk 12]
```diff
@@ -82,52 +84,70 @@ export function UndoModal({ isOpen, onClose, handleDelete }: UndoModalProps) {
...
 
-  if (isPending)
-    return (
-      <div className="flex w-full justify-center">
-        <Spinner />
-      </div>
+  const handlePreviewEditSubmit = (
+    payload: UploadCustomCodesPreviewItem | null
+  ) => {
+    if (payload === null) {
```
Comment (jakewheeler) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/Modals.tsx]:
```suggestion
    if (!payload) {
```

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx (hunk 7 above)
Comment (fzhao99) [reply client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]: works for me.

File: client/e2e/configuration.spec.ts (hunk 6 above)
Comment (fzhao99) [reply client/e2e/configuration.spec.ts:490]: hmm ok, will look into how complicated it is to fix it and make a ticket if it makes sense

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx (hunk 7 above)
Comment (jakewheeler) [reply client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]: #1466 🫡

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx (hunk 9 above)
Comment (fzhao99) [reply client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]: alright @jakewheeler I think by componentizing the CSV preview table into its own component and compressing the error states into one, was able to get the states down from 8 --> 5. Let me know if we feel it's more manageable now / if there are other ways you can think of simplifying

File: client/e2e/configuration.spec.ts (hunk 6 above)
Comment (fzhao99) [reply client/e2e/configuration.spec.ts:490]: Think these slate of changes may have resolved this? 
https://github.com/user-attachments/assets/20ff3e7c-85ca-4f89-87fd-74a34fd88064

Comment (jakewheeler) [reply client/e2e/configuration.spec.ts:490]: Nice 👀

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx [hunk 13]
```diff
@@ -127,6 +132,27 @@ describe('Custom codes upload', () => {
     checkCvxCode();
   });
 
+  test('successive row deletes reset to the instruction step', async () => {
```
Comment (fzhao99) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.test.tsx:135]: realized we had a side-effect state for "if all the rows are deleted, go back to the instructions", so added this extra test

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/Modals.tsx [hunk 14]
```diff
@@ -144,10 +143,10 @@ export function PreviewEditModal({
     !previewEditForm.system_key;
 
   return (
-    <Modal open={isOpen} onClose={closePreviewEditModal}>
+    <Modal open={true} onClose={closePreviewEditModal}>
```
Comment (fzhao99) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/Modals.tsx]: lmk if this way of driving the modal state is an anti pattern: it does make the type handling easier, but could also pass in a potential null state for the edit item and then handle the undefined case accordingly

Comment (jakewheeler) [reply client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/Modals.tsx]: Solely for the sake of readability I'd lean towards passing in an `open` indicator. It doesn't really change anything from a tech standpoint but I would find it easier to follow what's happening.

Defining something basic like this in `PreviewEditTable` and passing it along to the modal's `open` should do the job.
```ts
const isPreviewEditModalOpen = previewEditItem !== null;
```

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx [hunk 15]
```diff
@@ -556,99 +379,241 @@ export function ImportCustomCodes({
...
+function PreviewEditTable({
+  previewDisplayItems,
+  setPreviewItems,
+  codeSystems,
+  error,
+  setError,
+  exitPreviewStep,
+}: PreviewEditTableProps) {
+  const [previewEditItem, setPreviewEditItem] =
+    useState<UploadCustomCodesPreviewItem | null>(null);
```
Comment (jakewheeler) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]: What do you think about a slight name tweak here?

```suggestion
  const [itemBeingEdited, setItemBeingEdited] =
    useState<UploadCustomCodesPreviewItem | null>(null);
```

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx [hunk 16]
```diff
@@ -556,99 +379,241 @@ export function ImportCustomCodes({
...
+  const showToast = useToast();
+
+  const openPreviewEditModal = (editIndex: string) => {
+    const target = previewDisplayItems.find((i) => i.item.id === editIndex);
+    if (!target) return;
+    setPreviewEditItem({ ...target.item });
```
Comment (jakewheeler) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/ImportCustomCodes.tsx]:
```suggestion
    if (target) setPreviewEditItem({ ...target.item });
```

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/Modals.tsx [hunk 17]
```diff
@@ -82,52 +85,68 @@ export function UndoModal({ isOpen, onClose, handleDelete }: UndoModalProps) {
...
+    !previewEditForm.name ||
+    !previewEditForm.system_key;
 
   return (
-    <Modal open={isOpen} onClose={closePreviewEditModal}>
+    <Modal open={true} onClose={closePreviewEditModal}>
       <ModalHeader>
         <ModalTitle>
-          {previewEditForm.code ? `Edit ${previewEditForm.code}` : 'Edit code'}
+          {previewEditItem?.code ? `Edit ${previewEditItem.code}` : 'Edit code'}
```
Comment (jakewheeler) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/Modals.tsx]: Is this possible? It looks like `previewEditItem` can't be `null`.

File: client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/Modals.tsx [hunk 18]
```diff
@@ -136,17 +155,20 @@ export function PreviewEditModal({
             <Label>Code</Label>
             <TextInput
               type="text"
-              value={previewEditForm.code}
-              onChange={handlePreviewEditChange('code')}
+              value={previewEditForm?.code}
```
Comment (jakewheeler) [inline client/src/pages/Configurations/ConfigBuild/CustomCodes/CsvImport/Modals.tsx]: Same question about the optional chaining here.

File: refiner/app/api/v1/configurations/custom_codes.py [hunk 19]
```diff
@@ -171,40 +173,38 @@ class UploadCustomCodesPreviewResponse(BaseModel):
...
 
-    Returns:
-        UploadCustomCodesResponse
-    """
+def _validate_required_columns_or_raise(csv_reader: csv.DictReader[str]):
+    headers = set(csv_reader.fieldnames or [])
 
-    if body.filename and not body.filename.lower().endswith(".csv"):
+    has_system = "code_system" in (headers)
+    # maintain backwards compatibility with old template that used "code_number" as the column header
```
Comment (jakewheeler) [inline refiner/app/api/v1/configurations/custom_codes.py]: I don't have much of an issue with this but also right now will be the easiest time we'll ever have to get rid of the old naming 😆 It doesn't seem too risky to me to stop supporting `code_number` entirely right now.

Comment (fzhao99) [reply refiner/app/api/v1/configurations/custom_codes.py]: fair enough! removed

**Review by jakewheeler — APPROVED** [review]
🫡 🚀
